File: ios/GuideStreamTV/Services/WatchIntentLogger.swift
//
//  WatchIntentLogger.swift
//  GuideStreamTV
//

import Foundation
import Supabase

enum IntentEventType: String {
    case cardTapped = "card_tapped"
    case deeplinkFired = "deeplink_fired"
    case notificationOpened = "notification_opened"
    case searchQuery = "search_query"
    case trailerWatched = "trailer_watched"
    case trailerSkipped = "trailer_skipped"
    case streamAdded = "stream_added"
    case streamRemoved = "stream_removed"
    case bingeAlertOpened = "binge_alert_opened"
    case askStreamQuery = "ask_stream_query"
    case playOnDeviceChosen = "play_on_device_chosen"
    case episodeDetailViewed = "episode_detail_viewed"
    case continueWatching = "continue_watching_tapped"
    case widgetSetupTapped = "widget_setup_tapped"
    case affiliateLinkTapped = "affiliate_link_tapped"
    case sponsoredReelViewed = "sponsored_reel_viewed"
    case sponsoredReelTapped = "sponsored_reel_tapped"
    case adImpression = "ad_impression"
    case trailerViewed = "trailer_viewed"
    case trailerLiked = "trailer_liked"
    case notifyReleaseTapped = "notify_release_tapped"
    case commentsOpened = "comments_opened"
    case muteToggled = "mute_toggled"
    case watchlistAdded = "watchlist_added"
    case watchlistRemoved = "watchlist_removed"
    case watchedToggled = "watched_toggled"

    // Lifecycle events — fire whether signed in or not so we capture every device.
    case sessionStarted = "session_started"
    case authSignedIn = "auth_signed_in"
    case guestStarted = "guest_started"
    case onboardingCompleted = "onboarding_completed"
    case serviceSelected = "service_selected"
    case appOpened = "app_opened"
}

/// Captured error entry — surfaced in the in-app diagnostics screen so the
/// user can see *exactly* why a Supabase write failed (RLS, missing table,
/// network, etc.) instead of guessing.
nonisolated struct LoggerError: Identifiable, Sendable {
    let id: UUID = UUID()
    let timestamp: Date
    let eventType: String
    let message: String
}

/// Fire-and-forget logger that writes a row to `watch_intent_events` for every
/// meaningful user action — **whether the user is signed in or a guest**.
///
/// Every payload includes:
/// - `event_type` — the action being recorded.
/// - `user_id` — the Supabase auth user id when signed in (omitted for guests).
/// - `device_id` — stable per-device UUID, populated for every user. Lets us
///    attribute guest activity to a real device install.
/// - `metadata.device_id` / `metadata.is_guest` — also mirrored into the JSON
///    metadata blob in case the table doesn't yet have a top-level `device_id`
///    column. Mirroring guarantees the data is captured either way.
///
/// Errors are logged to the console *verbosely* and kept in a ring buffer so
/// the `SupabaseDiagnosticsView` can show the last few failures.
@MainActor
final class WatchIntentLogger {
    static let shared = WatchIntentLogger()
    private init() {}

    /// Last N error messages. Kept on the main actor so SwiftUI views can
    /// observe it directly via `@State`/`@Observable` patterns.
    private(set) var recentErrors: [LoggerError] = []

    /// Total number of insert attempts (success or failure) since launch.
    private(set) var totalAttempts: Int = 0
    /// Total number of successful inserts since launch.
    private(set) var totalSuccesses: Int = 0

    private let maxErrors: Int = 20

    func log(
        eventType: IntentEventType,
        titleId: String? = nil,
        platformId: String? = nil,
        metadata: [String: Any]? = nil,
        watchDurationSeconds: Double? = nil
    ) {
        // Snapshot state synchronously on the main actor before hopping to a
        // background Task — keeps everything Sendable.
        let userId = AuthViewModel.shared.currentUser?.id.uuidString
        let isGuest = AuthViewModel.shared.isGuest && userId == nil
        let deviceId = DeviceIdentity.shared.deviceId
        let event = eventType.rawValue
        let titleIdCopy = titleId
        let platformIdCopy = platformId

        // Build merged metadata up front (still on MainActor — no Sendable
        // captures of [String: Any] across the Task boundary).
        var mergedMeta: [String: Any] = metadata ?? [:]
        mergedMeta["device_id"] = deviceId
        mergedMeta["is_guest"] = isGuest
        mergedMeta["is_authenticated"] = userId != nil
        if let duration = watchDurationSeconds {
            mergedMeta["watch_duration_seconds"] = duration
        }
        let metadataJSON: AnyJSON? = Self.toAnyJSON(mergedMeta)

        totalAttempts += 1

        Task { [weak self] in
            var payload: [String: AnyJSON] = [
                "event_type": .string(event),
                "device_id": .string(deviceId)
            ]
            if let userId { payload["user_id"] = .string(userId) }
            if let titleIdCopy { payload["title_id"] = .string(titleIdCopy) }
            if let platformIdCopy { payload["platform_id"] = .string(platformIdCopy) }
            if let metadataJSON { payload["metadata"] = metadataJSON }

            do {
                try await SupabaseManager.shared.client
                    .from("watch_intent_events")
                    .insert(payload)
                    .execute()
                await self?.recordSuccess()
                // Notify observers so Profile stats refresh reactively.
                await self?.postStatsNotification()
            } catch {
                // If the table is missing a top-level `device_id` column,
                // Postgres will reject the insert with a 400/column error.
                // Retry once with `device_id` dropped — the metadata still
                // carries it so the data isn't lost.
                let message = Self.describe(error)
                let columnIssue = message.localizedCaseInsensitiveContains("device_id")
                    && (message.localizedCaseInsensitiveContains("column")
                        || message.localizedCaseInsensitiveContains("schema")
                        || message.localizedCaseInsensitiveContains("could not find"))
                if columnIssue {
                    var fallback = payload
                    fallback.removeValue(forKey: "device_id")
                    do {
                        try await SupabaseManager.shared.client
                            .from("watch_intent_events")
                            .insert(fallback)
                            .execute()
                        await self?.recordSuccess()
                        await self?.postStatsNotification()
                        return
                    } catch {
                        self?.recordError(event: event, error: error)
                        return
                    }
                }
                self?.recordError(event: event, error: error)
            }
        }
    }

    private func recordSuccess() {
        totalSuccesses += 1
    }

    /// Posts a notification so `ProfileStatsService` knows new engagement
    /// data is available and can refresh without a manual pull.
    private func postStatsNotification() {
        NotificationCenter.default.post(
            name: Notification.Name("ProfileStatsNeedsRefresh"),
            object: nil
        )
    }

    private func recordError(event: String, error: Error) {
        let message = Self.describe(error)
        print("[WatchIntent ERROR] \(event): \(message)")
        let entry = LoggerError(timestamp: Date(), eventType: event, message: message)
        recentErrors.insert(entry, at: 0)
        if recentErrors.count > maxErrors {
            recentErrors.removeLast(recentErrors.count - maxErrors)
        }
    }

    /// Manually fire a test event from the diagnostics screen — returns the
    /// resulting error message (if any) so the UI can render it inline.
    func logTestEvent() async -> String? {
        let deviceId = DeviceIdentity.shared.deviceId
        let userId = AuthViewModel.shared.currentUser?.id.uuidString
        let isGuest = AuthViewModel.shared.isGuest && userId == nil
        var payload: [String: AnyJSON] = [
            "event_type": .string("diagnostic_ping"),
            "device_id": .string(deviceId)
        ]
        if let userId { payload["user_id"] = .string(userId) }
        let meta: [String: Any] = [
            "device_id": deviceId,
            "is_guest": isGuest,
            "is_authenticated": userId != nil,
            "source": "diagnostics"
        ]
        if let metaJSON = Self.toAnyJSON(meta) { payload["metadata"] = metaJSON }

        totalAttempts += 1
        do {
            try await SupabaseManager.shared.client
                .from("watch_intent_events")
                .insert(payload)
                .execute()
            totalSuccesses += 1
            return nil
        } catch {
            let message = Self.describe(error)
            recordError(event: "diagnostic_ping", error: error)
            return message
        }
    }

    /// Lowercases and dashes a free-form title into a stable id slug, e.g.
    /// "Stranger Things" → "tt-stranger-things".
    static func titleSlug(_ title: String) -> String {
        let lower = title.lowercased()
        let allowed = lower.map { ch -> Character in
            if ch.isLetter || ch.isNumber { return ch }
            return "-"
        }
        var slug = String(allowed)
        while slug.contains("--") { slug = slug.replacingOccurrences(of: "--", with: "-") }
        slug = slug.trimmingCharacters(in: CharacterSet(charactersIn: "-"))
        return "tt-\(slug)"
    }

    // MARK: - AnyJSON conversion

    nonisolated private static func toAnyJSON(_ value: Any) -> AnyJSON? {
        switch value {
        case let s as String: return .string(s)
        case let b as Bool: return .bool(b)
        case let i as Int: return .integer(i)
        case let d as Double: return .double(d)
        case let f as CGFloat: return .double(Double(f))
        case let arr as [Any]:
            return .array(arr.compactMap { toAnyJSON($0) })
        case let dict as [String: Any]:
            var out: [String: AnyJSON] = [:]
            for (k, v) in dict {
                if let j = toAnyJSON(v) { out[k] = j }
            }
            return .object(out)
        default:
            return .string(String(describing: value))
        }
    }

    /// Produce a verbose human-readable description for an error, including
    /// Postgres response details when present.
    nonisolated private static func describe(_ error: Error) -> String {
        let ns = error as NSError
        var parts: [String] = [ns.localizedDescription]
        if let underlying = ns.userInfo[NSUnderlyingErrorKey] as? Error {
            parts.append("underlying: \((underlying as NSError).localizedDescription)")
        }
        for key in ["message", "hint", "details", "code"] {
            if let value = ns.userInfo[key] as? String, !value.isEmpty {
                parts.append("\(key)=\(value)")
            }
        }
        return parts.joined(separator: " | ")
    }
}

File: android/app/src/main/java/com/rork/guidestreamtvandroid/data/repository/WatchIntentLogger.kt
package com.rork.guidestreamtvandroid.data.repository

import android.content.Context
import android.util.Log
import com.rork.guidestreamtvandroid.data.local.DeviceIdentity
import com.rork.guidestreamtvandroid.data.remote.SupabaseManager
import io.github.jan.supabase.postgrest.postgrest
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * Fire-and-forget analytics logger — mirrors iOS WatchIntentLogger.swift.
 * Writes a row to `watch_intent_events` for every meaningful user action,
 * whether the user is signed in or a guest.
 */
class WatchIntentLogger private constructor(context: Context) {

    enum class IntentEventType(val value: String) {
        CARD_TAPPED("card_tapped"),
        DEEPLINK_FIRED("deeplink_fired"),
        NOTIFICATION_OPENED("notification_opened"),
        SEARCH_QUERY("search_query"),
        TRAILER_WATCHED("trailer_watched"),
        TRAILER_SKIPPED("trailer_skipped"),
        STREAM_ADDED("stream_added"),
        STREAM_REMOVED("stream_removed"),
        BINGE_ALERT_OPENED("binge_alert_opened"),
        ASK_STREAM_QUERY("ask_stream_query"),
        PLAY_ON_DEVICE_CHOSEN("play_on_device_chosen"),
        EPISODE_DETAIL_VIEWED("episode_detail_viewed"),
        CONTINUE_WATCHING("continue_watching_tapped"),
        WIDGET_SETUP_TAPPED("widget_setup_tapped"),
        AFFILIATE_LINK_TAPPED("affiliate_link_tapped"),
        SPONSORED_REEL_VIEWED("sponsored_reel_viewed"),
        SPONSORED_REEL_TAPPED("sponsored_reel_tapped"),
        AD_IMPRESSION("ad_impression"),
        TRAILER_VIEWED("trailer_viewed"),
        TRAILER_LIKED("trailer_liked"),
        NOTIFY_RELEASE_TAPPED("notify_release_tapped"),
        COMMENTS_OPENED("comments_opened"),
        MUTE_TOGGLED("mute_toggled"),
        WATCHLIST_ADDED("watchlist_added"),
        WATCHLIST_REMOVED("watchlist_removed"),
        SESSION_STARTED("session_started"),
        AUTH_SIGNED_IN("auth_signed_in"),
        GUEST_STARTED("guest_started"),
        ONBOARDING_COMPLETED("onboarding_completed"),
        SERVICE_SELECTED("service_selected"),
        APP_OPENED("app_opened"),
        WATCHED_TOGGLED("watched_toggled"),
    }

    data class LoggerError(
        val timestampMs: Long,
        val eventType: String,
        val message: String,
    )

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    val recentErrors = mutableListOf<LoggerError>()
    var totalAttempts: Int = 0
        private set
    var totalSuccesses: Int = 0
        private set

    companion object {
        private const val TAG = "WatchIntent"
        private const val MAX_ERRORS = 20

        @Volatile private var instance: WatchIntentLogger? = null
        fun init(context: Context): WatchIntentLogger =
            instance ?: synchronized(this) {
                instance ?: WatchIntentLogger(context.applicationContext).also { instance = it }
            }
        fun get(): WatchIntentLogger =
            instance ?: error("WatchIntentLogger not initialized")
    }

    fun log(
        eventType: IntentEventType,
        titleId: String? = null,
        platformId: String? = null,
        metadata: Map<String, Any?> = emptyMap(),
        watchDurationSeconds: Double? = null,
    ) {
        val auth = AuthViewModel.get()
        val userId = auth.currentUserId
        val isGuest = auth.isGuest.value && userId == null
        val deviceId = DeviceIdentity.get().deviceId
        val environment = DeviceIdentity.get().environment
        val event = eventType.value

        val mergedMeta = metadata.toMutableMap()
        mergedMeta["device_id"] = deviceId
        mergedMeta["environment"] = environment
        mergedMeta["is_guest"] = isGuest
        mergedMeta["is_authenticated"] = userId != null
        if (watchDurationSeconds != null) {
            mergedMeta["watch_duration_seconds"] = watchDurationSeconds
        }

        val metadataJson = toJsonObject(mergedMeta)

        totalAttempts += 1

        scope.launch {
            val payload = buildJsonObject {
                put("event_type", event)
                put("device_id", deviceId)
                put("environment", environment)
                if (userId != null) put("user_id", userId)
                if (titleId != null) put("title_id", titleId)
                if (platformId != null) put("platform_id", platformId)
                put("metadata", metadataJson)
            }
            try {
                SupabaseManager.client.postgrest
                    .from("watch_intent_events")
                    .insert(payload)
                synchronized(recentErrors) { totalSuccesses += 1 }
            } catch (e: Throwable) {
                if (e is CancellationException) throw e
                val message = e.message ?: "unknown"
                val columnIssue = message.contains("device_id", ignoreCase = true) &&
                    (message.contains("column", ignoreCase = true) ||
                        message.contains("schema", ignoreCase = true) ||
                        message.contains("could not find", ignoreCase = true))
                if (columnIssue) {
                    try {
                        val fallback = buildJsonObject {
                            put("event_type", event)
                            put("environment", environment)
                            if (userId != null) put("user_id", userId)
                            if (titleId != null) put("title_id", titleId)
                            if (platformId != null) put("platform_id", platformId)
                            put("metadata", metadataJson)
                        }
                        SupabaseManager.client.postgrest
                            .from("watch_intent_events")
                            .insert(fallback)
                        synchronized(recentErrors) { totalSuccesses += 1 }
                        return@launch
                    } catch (e2: Throwable) {
                        if (e2 is CancellationException) throw e2
                        recordError(event, e2.message ?: "unknown")
                        return@launch
                    }
                }
                recordError(event, message)
            }
        }
    }

    private fun recordError(event: String, message: String) {
        Log.e(TAG, "$event: $message")
        synchronized(recentErrors) {
            recentErrors.add(0, LoggerError(System.currentTimeMillis(), event, message))
            if (recentErrors.size > MAX_ERRORS) {
                recentErrors.subList(MAX_ERRORS, recentErrors.size).clear()
            }
        }
    }

    /**
     * Lowercases and dashes a free-form title into a stable id slug,
     * e.g. "Stranger Things" → "tt-stranger-things".
     */
    fun titleSlug(title: String): String {
        val lower = title.lowercase()
        val slug = lower.map { ch ->
            if (ch.isLetterOrDigit()) ch else '-'
        }.joinToString("")
            .replace("--", "-")
            .trim('-')
        return "tt-$slug"
    }

    private fun toJsonObject(map: Map<String, Any?>): JsonObject = buildJsonObject {
        for ((key, value) in map) {
            when (value) {
                null -> {}
                is String -> put(key, value)
                is Boolean -> put(key, value)
                is Int -> put(key, value)
                is Long -> put(key, value)
                is Double -> put(key, value)
                is Float -> put(key, value.toDouble())
                is List<*> -> put(key, toJsonArray(value))
                is Map<*, *> -> {
                    @Suppress("UNCHECKED_CAST")
                    put(key, toJsonObject(value as Map<String, Any?>))
                }
                else -> put(key, value.toString())
            }
        }
    }

    private fun toJsonArray(list: List<*>): JsonArray = JsonArray(
        list.mapNotNull { item ->
            when (item) {
                is String -> JsonPrimitive(item)
                is Boolean -> JsonPrimitive(item)
                is Int -> JsonPrimitive(item)
                is Long -> JsonPrimitive(item)
                is Double -> JsonPrimitive(item)
                is Map<*, *> -> {
                    @Suppress("UNCHECKED_CAST")
                    toJsonObject(item as Map<String, Any?>)
                }
                is List<*> -> toJsonArray(item)
                else -> null
            }
        },
    )
}
